proposals/improved-interpolated-strings.md

```diff
@@ -0,0 +1,436 @@
+receiver of the method, that will allow things like logging frameworks to avoid allocating strings that will never be needed, and giving C# users familiar,
+convenient interpolation syntax.
+
+## Detailed Design
+
+### The builder pattern
+
+We introduce a new builder pattern that can represent an interpolated string passed as an argument to a method. The simple English of the pattern is as follows:
+
+When an _interpolated\_string\_expression_ is passed as an argument to a method, we look at the type of the parameter. If the parameter type has a static method
+`GetInterpolatedStringBuilder` that can invoked with 2 int parameters, `baseLength` and `formatHoleCount`, optionally takes a parameter the receiver is convertible to,
+and has an out parameter of the type of original method's parameter and that type has instance `TryFormat` methods can be invoked for every part of the interpolated
+string, then we lower the interpolation using that, instead of into a traditional call to `string.Format(formatStr, args)`. A more concrete example is helpful for
+picturing this:
+
+```cs
+// The builder that will actually "build" the interpolated string"
+public ref struct TraceLoggerParamsBuilder
+{
+    public static void GetInterpolatedStringBuilder(int baseLength, int formatHoleCount, Logger logger, out TraceLoggerParamsBuilder builder)
+    {
+        builder = TraceLoggerParamsBuilder(baseLength, formatHoleCount, logger.EnabledLevel);
+    }
+
+    // Storage for the built-up string
+
+    private bool _logLevelEnabled;
+
+    private TraceLoggerParamsBuilder(int baseLength, int formatHoleCount, bool logLevelEnabled)
+    {
+        // Initialization logic
+        _logLevelEnabled = logLevelEnabled
+    }
+
+    public bool TryFormat(string s)
+    {
+        if (!_logLevelEnabled) return false;
+
+        // Store and format part as required
+        return true;
+    }
+
+    public bool TryFormat<T>(T t)
+    {
+        if (!_logLevelEnabled) return false;
+
+        // Store and format part as required
+        return true;
+    }
+}
+
+// The logger class. The user has an instance of this, accesses it via static state, or some other access
+// mechanism
+public class Logger
+{
+    // Initialization code omitted
+    public LogLevel EnabledLevel;
+
+    public void LogTrace(TraceLoggerParamsBuilder builder)
+    {
+
+        // Impl of logging
+    }
+}
+
+Logger logger = GetLogger(LogLevel.Info);
+
+// Given the above definitions, usage looks like this:
+logger.LogTrace($"{"this"} will never be printed because info is < trace!");
+
+// This is converted to:
+var receiverTemp = logger;
+TraceLoggerParamsBuilder.GetInterpolatedStringBuilder(baseLength: 47, formatHoleCount: 1, receiverTemp, out var builder);
+_ = builder.TryFormat("this") && builder.TryFormat(" will never be printed because info is < trace!");
+receiverTemp.Log(builder);
```

Comment:
Was this supposed to be ?:

```suggestion
receiverTemp.LogTrace(builder);
```

Reply:
Yep, thanks. Missed this when converted from an earlier version.